sky/utils/vpn_utils.py

```diff
@@ -0,0 +1,309 @@
+                    'must be set to enable the Tailscale API.')
+
+    @staticmethod
+    def from_env_vars() -> 'TailscaleVPN':
+        # Parse Tailscale auth key from environment variable.
+        # This is required for all Tailscale operations.
+        auth_key = os.environ.get('TAILSCALE_AUTH_KEY')
+        if auth_key is None:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError('TAILSCALE_AUTH_KEY is not set.')
+
+        # Parse Tailscale API key and network name
+        api_key = os.environ.get('TAILSCALE_API_KEY')
+        network_name = os.environ.get('TAILSCALE_NETWORK_NAME')
+        enable_api = True
+        if not api_key or not network_name:
+            enable_api = False
+
+        return TailscaleVPN(auth_key, api_key, network_name, enable_api)
+
+    def get_setup_command(self, hostname: str) -> str:
+        return self._SETUP_COMMAND.format(tailscale_auth_key=self._auth_key,
+                                          hostname=hostname)
+
+    def get_setup_env_vars(self) -> Dict[str, str]:
+        env_config = {
+            'TAILSCALE_AUTH_KEY': self._auth_key,
+        }
+        if self._enable_api:
+            assert self._api_key and self._network_name
+            env_config['TAILSCALE_API_KEY'] = self._api_key
+            env_config['TAILSCALE_NETWORK_NAME'] = self._network_name
+        return env_config
+
+    def _get_auth_headers(self) -> Dict[str, str]:
+        """Get the authentication headers for the Tailscale API."""
+        return {
+            'Authorization': f'Bearer {self._api_key}',
+        }
+
+    def _get_device_id_from_hostname(self, hostname: str) -> Optional[str]:
+        """Get the node ID from the hostname."""
+        url_to_query = ('https://api.tailscale.com/api/v2/tailnet/'
+                        f'{self._network_name}/devices')
+        resp = requests.get(url_to_query, headers=self._get_auth_headers())
+        all_devices_in_network = resp.json().get('devices', [])
+        for device_info in all_devices_in_network:
+            if device_info.get('hostname') == hostname:
+                return device_info.get('id')
+        return None
+
+    def get_private_ip(self, hostname: str) -> str:
+        """Get the private IP address from the hostname."""
+        if self._enable_api:
+            device_id = self._get_device_id_from_hostname(hostname)
+            url_to_query = f'https://api.tailscale.com/api/v2/device/{device_id}'
+            resp = requests.get(url_to_query, headers=self._get_auth_headers())
+            return resp.json().get('addresses', [])[0]
+        # If the API is not enabled, we need to query the IP address locally.
+        # However, since currently self._enable_api is always True, this code
+        # will not be executed.
+        query_cmd = f'tailscale ip -4 {hostname}'
+        rc, stdout, stderr = subprocess_utils.run_with_retries(
+            query_cmd,
+            max_retry=1000,
```

Comment:
lets use a reasonable retry times (like 5). in the previous implementation, we also need to wait the instance to boot up, that is the reason to use 1000 here. Now we are calling this after the instance is ready so this should works immediately?